src/serialize.ts

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+
+"use strict";
+
+import utils = require("./utils");
+import kernel = require("./kernel");
+
+import IKernelMsg = kernel.IKernelMsg;
+
+
+function _deserializeArrayBuffer(buf: ArrayBuffer): IKernelMsg {
+  var data = new DataView(buf);
+  // read the header: 1 + nbufs 32b integers
+  var nbufs = data.getUint32(0);
+  var offsets: number[] = [];
+  var i: number;
+  for (i = 1; i <= nbufs; i++) {
+    offsets.push(data.getUint32(i * 4));
+  }
+  var json_bytes = new Uint8Array(buf.slice(offsets[0], offsets[1]));
+  var msg = JSON.parse(
+    (new TextDecoder('utf8')).decode(json_bytes)
+    );
+  // the remaining chunks are stored as DataViews in msg.buffers
+  msg.buffers = [];
+  var start: number, stop: number;
+  for (i = 1; i < nbufs; i++) {
+    start = offsets[i];
+    stop = offsets[i + 1] || buf.byteLength;
+    msg.buffers.push(new DataView(buf.slice(start, stop)));
+  }
+  return msg;
+};
+
+
+/**
+ * deserialize the binary message format
+ * callback will be called with a message whose buffers attribute
+ * will be an array of DataViews.
+ */
+function _deserializeBinary(data: Blob | ArrayBuffer): IKernelMsg | Promise<IKernelMsg> {
```

Comment:
Maybe always make this a promise return value? Then the caller can handle it consistently. You can return an already resolved promise via `Promise.resolve(value)`.